In [0]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

In [0]:
from fastai.text import*
from fastai.imports import *

In [10]:
# Run this cell and select the kaggle.json file downloaded
# from the Kaggle account settings page.
from google.colab import files
files.upload()

Saving questions.csv to questions (1).csv


{'questions.csv': b',,,,,,\r\n2,\xe0\xa4\x95\xe0\xa5\x8d\xe0\xa4\xaf\xe0\xa4\xbe \xe0\xa4\xac\xe0\xa5\x80\xe0\xa4\x9c\xe0\xa5\x87\xe0\xa4\xaa\xe0\xa5\x80 \xe0\xa4\x95\xe0\xa5\x8b \xe0\xa4\xaa\xe0\xa5\x8d\xe0\xa4\xb0\xe0\xa4\xbf\xe0\xa4\xaf\xe0\xa4\x82\xe0\xa4\x95\xe0\xa4\xbe \xe0\xa4\x95\xe0\xa5\x80 \xe0\xa4\x9f\xe0\xa4\x95\xe0\xa5\x8d\xe0\xa4\x95\xe0\xa4\xb0 \xe0\xa4\xae\xe0\xa5\x87\xe0\xa4\x82 \xe0\xa4\xb5\xe0\xa4\xb0\xe0\xa5\x81\xe0\xa4\xa3 \xe0\xa4\x97\xe0\xa4\xbe\xe0\xa4\x82\xe0\xa4\xa7\xe0\xa5\x80 \xe0\xa4\x95\xe0\xa5\x8b \xe0\xa4\x96\xe0\xa4\xa1\xe0\xa4\xbc\xe0\xa4\xbe \xe0\xa4\x95\xe0\xa4\xb0\xe0\xa4\xa8\xe0\xa4\xbe \xe0\xa4\x9a\xe0\xa4\xbe\xe0\xa4\xb9\xe0\xa4\xbf\xe0\xa4\x8f \xe0\xa4\x94\xe0\xa4\xb0 \xe0\xa4\x95\xe0\xa5\x8d\xe0\xa4\xaf\xe0\xa5\x8b\xe0\xa4\x82 ?,,,,,\r\n2,Narander modi ji ka speech aapko ksa lgta h ?,,,,,\r\n2,\xe0\xa4\x95\xe0\xa5\x8d\xe0\xa4\xaf\xe0\xa4\xbe \xe0\xa4\xae\xe0\xa4\xae\xe0\xa4\xa4\xe0\xa4\xbe \xe0\xa4\xac\xe0\xa4\xa8\xe0\xa4\xb0\xe0\xa5\x8d\xe0\xa

# preparing the data

 we have totally 376 questions labelled as category 2 and 3. And 2 is for politics , 3 is for movies

In [11]:
df = pd.read_csv('questions.csv',encoding='utf-8')
df.head()

Unnamed: 0  ... Unnamed: 6
0         2.0  ...        NaN
1         2.0  ...        NaN
2         2.0  ...        NaN
3         2.0  ...        NaN
4         2.0  ...        NaN

[5 rows x 7 columns]

# data cleaning

In [0]:
df = df.drop(df[['Unnamed: 2','Unnamed: 3','Unnamed: 4','Unnamed: 5','Unnamed: 6']], axis=1)

In [13]:
df.rename(columns={'Unnamed: 1':'Query','Unnamed: 0':'Category'}, inplace=True)
df

Category                                              Query
0          2.0  क्या बीजेपी को प्रियंका की टक्कर में वरुण गांध...
1          2.0      Narander modi ji ka speech aapko ksa lgta h ?
2          2.0  क्या ममता बनर्जी को धरने में बैठने से कोई फायद...
3          2.0       भारत की राजनीति में सबसे घटिया आदमी कौन है ?
4          2.0  अरविंद केजरीवाल को प्रधानमंत्री मोदी इतने बुरे...
5          2.0  पीएम मोदी राफेल सौदे से जुड़े विपक्ष के सवालों...
6          2.0  यदि सपा बसपा गठबंधन में कॉन्ग्रेस भी शामिल हो ...
7          2.0  हिंदू-मुसलमान विवाद से किसी देश को किस प्रकार ...
8          2.0  कौन सा भाषा है जो कि सब से ज्यादा देशों में बो...
9          2.0  गौरव जी आप यह बताइए कि आप ने बीजेपी की पार्टी ...
10         2.0  भारत की राजधानी दिल्ली को पूर्ण राज्य का दर्जा...
11         2.0  हमारे भारत के कुछ जवान शहीद हो गए हैं और तो आप...
12         2.0  अधिकांशतः भ्रष्ट नेता इंडिया में ही क्यों होते...
13         2.0  जम्मू कश्मीर के पुलवामा में हुए आतंकी हमले पर ...
14         2.0  आपके अनुसार भारतीय राजनीति में किस बड़े सुधार ...
15         2.0  कश्मीर की कुछ लोग पाकिस्तान का साथ देते हैं अब...
16         2.0  भारत मे जो पाकिस्तान के चहिते है उनको सरकार कु...
17         2.0  100 में से 90 प्रतिशत मुस्लिम ही क्यों आतंकवाद...
18         2.0  पुलवामा में हुए आतंकवादी हमले के 4 दिन बीत गए ...
19         2.0  आजकल भारत जैसे देश में बहुत सारे मुद्दे हैं उस...
20         2.0  भारत में कुछ जगहों से कश्मीरियों को निकाला जा ...
21         2.0  लोकसभा चुनाव के मद्देनजर भाजपा और एआईडीएमके के...
22         2.0  भारत का जो कानून है वह लचीला हैं ? अगर हां तो ...
23         2.0  भारत में लोगों के पुलिस के साथ अच्छे संबंध चलत...
24         2.0  कुछ बदनाम किस्से सऊदी के रॉयल फैमिली के बारे म...
25         2.0  अगर स्मृति ईरानी राहुल गांधी के खिलाफ 2019 में...
26         2.0  निरंतर हो रही फौजियों की मौत को लेकर क्या सरका...
27         2.0  भारत का युवा वर्ग आज किस दिशा की और जा रहा है ...
28         2.0     क्या बेरोजगार अभी एक चुनावी मुद्दा बन चुका है?
29         2.0  हम हमारे देश की अर्थव्यवस्था को किस प्रकार मजब...
...        ...                                                ...
973        NaN                                                NaN
974        NaN                                                NaN
975        NaN                                                NaN
976        NaN                                                NaN
977        NaN                                                NaN
978        NaN                                                NaN
979        NaN                                                NaN
980        NaN                                                NaN
981        NaN                                                NaN
982        NaN                                                NaN
983        NaN                                                NaN
984        NaN                                                NaN
985        NaN                                                NaN
986        NaN                                                NaN
987        NaN                                                NaN
988        NaN                                                NaN
989        NaN                                                NaN
990        NaN                                                NaN
991        NaN                                                NaN
992        NaN                                                NaN
993        NaN                                                NaN
994        NaN                                                NaN
995        NaN                                                NaN
996        NaN                                                NaN
997        NaN                                                NaN
998        NaN                                                NaN
999        NaN                                                NaN
1000       NaN                                                NaN
1001       NaN                                

missing values

In [14]:
df.isnull().sum()

Category    628
Query       628
dtype: int64

In [0]:
 df['Category'].replace([2.0,3.0],['2','3'], inplace=True)

In [16]:
df.head()

Category                                              Query
0        2  क्या बीजेपी को प्रियंका की टक्कर में वरुण गांध...
1        2      Narander modi ji ka speech aapko ksa lgta h ?
2        2  क्या ममता बनर्जी को धरने में बैठने से कोई फायद...
3        2       भारत की राजनीति में सबसे घटिया आदमी कौन है ?
4        2  अरविंद केजरीवाल को प्रधानमंत्री मोदी इतने बुरे...

## **dropping missing values**

In [17]:
df=df.dropna(subset = ['Category', 'Query'])
df.head(3)

Category                                              Query
0        2  क्या बीजेपी को प्रियंका की टक्कर में वरुण गांध...
1        2      Narander modi ji ka speech aapko ksa lgta h ?
2        2  क्या ममता बनर्जी को धरने में बैठने से कोई फायद...

In [18]:
df.isnull().sum()

Category    0
Query       0
dtype: int64

In [19]:
df['Query'][3]

'भारत की राजनीति में सबसे घटिया आदमी कौन है ?'

# Tokenization

The first step of processing we make the texts go through is to split the raw sentences into words, or more exactly tokens. The easiest way to do this would be to split the string on spaces, but we can be smarter:

- we need to take care of punctuation
- some words are contractions of two different words, like isn't or don't

To see what the tokenizer had done behind the scenes, let's have a look at a few texts in a batch.

In [20]:
# import nltk
# nltk.download('stopwords')

# from nltk.corpus import stopwords 
# # stop_words
# stop_words = stopwords.words('english')
# stop_words
from google.colab import files
files.upload()

Saving hindi stopwords.xlsx to hindi stopwords (1).xlsx


{'hindi stopwords.xlsx': b'PK\x03\x04\x14\x00\x06\x00\x08\x00\x00\x00!\x00A7\x82\xcfn\x01\x00\x00\x04\x05\x00\x00\x13\x00\x08\x02[Content_Types].xml \xa2\x04\x02(\xa0\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x

In [21]:
stop_words=pd.read_excel('hindi stopwords.xlsx')
print (str(stop_words))

      के
0     का
1     एक
2    में
3     की
4     है
5     यह
6     और
7     से
8    हैं
9     को
10    पर
11    इस
12  होता
13    कि
14    जो
15    कर
16    मे
17   गया
18  करने
19  किया
20  लिये
21  अपने
22    ने
23   बनी
24  नहीं
25    तो
26    ही
27    या
28   एवं
29  दिया
..   ...
66    जा
67    ना
68   इसे
69  कहते
70    जब
71  होते
72   कोई
73   हुए
74     व
75     न
76   अभी
77  जैसे
78   सभी
79  करता
80  उनकी
81   तरह
82    उस
83   आदि
84   कुल
85    एस
86   रहा
87  इसकी
88  सकता
89   रहे
90  उनका
91   इसी
92  रखें
93  अपना
94    पे
95  उसके

[96 rows x 1 columns]


In [0]:
# detokenized_docs=[]
# t = ' '.join(tokenized_doc[1]) 
# detokenized_docs.append(t) 
# detokenized_docs

In [0]:
# tokenization 
tokenized_doc = df['Query'].apply(lambda x: x.split())
# tokenized_doc



In [24]:
# remove stop-words 
tokenized_doc = tokenized_doc.apply(lambda x: [item for item in x if item not in str(stop_words)])
tokenized_doc

0      [क्या, बीजेपी, प्रियंका, टक्कर, वरुण, गांधी, ख...
1      [Narander, modi, ji, ka, speech, aapko, ksa, l...
2      [क्या, ममता, बनर्जी, धरने, बैठने, फायदा, प्राप...
3             [भारत, राजनीति, सबसे, घटिया, आदमी, कौन, ?]
4      [अरविंद, केजरीवाल, प्रधानमंत्री, मोदी, इतने, ब...
5      [पीएम, मोदी, राफेल, सौदे, जुड़े, विपक्ष, सवालो...
6      [यदि, सपा, बसपा, गठबंधन, कॉन्ग्रेस, शामिल, जाए...
7      [हिंदू-मुसलमान, विवाद, किसी, देश, किस, प्रकार,...
8      [कौन, सा, भाषा, सब, ज्यादा, देशों, बोला, जाता,...
9      [गौरव, जी, आप, बताइए, आप, बीजेपी, पार्टी, क्यो...
10     [भारत, राजधानी, दिल्ली, पूर्ण, राज्य, दर्जा, म...
11     [हमारे, भारत, कुछ, जवान, शहीद, गए, आप, लोग, अप...
12        [अधिकांशतः, भ्रष्ट, नेता, इंडिया, क्यों, हैं?]
13     [जम्मू, कश्मीर, पुलवामा, आतंकी, हमले, पाकिस्ता...
14     [आपके, अनुसार, भारतीय, राजनीति, किस, बड़े, सुध...
15     [कश्मीर, कुछ, लोग, पाकिस्तान, साथ, देते, अब, भ...
16     [भारत, पाकिस्तान, चहिते, उनको, सरकार, कुछ, सजा...
17              [100, प्रतिशत, 

In [25]:
# de-tokenization 
detokenized_doc = [] 
for i in range(len(df)): 
    t = ' '.join(tokenized_doc[i]) 
    detokenized_doc.append(t) 

df['Query'] = detokenized_doc

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [26]:
df

Category                                              Query
0          2  क्या बीजेपी प्रियंका टक्कर वरुण गांधी खड़ा करन...
1          2      Narander modi ji ka speech aapko ksa lgta h ?
2          2     क्या ममता बनर्जी धरने बैठने फायदा प्राप्त होगा
3          2                 भारत राजनीति सबसे घटिया आदमी कौन ?
4          2  अरविंद केजरीवाल प्रधानमंत्री मोदी इतने बुरे क्...
5          2  पीएम मोदी राफेल सौदे जुड़े विपक्ष सवालों जवाब ...
6          2  यदि सपा बसपा गठबंधन कॉन्ग्रेस शामिल जाए क्या भ...
7          2  हिंदू-मुसलमान विवाद किसी देश किस प्रकार हानि ह...
8          2          कौन सा भाषा सब ज्यादा देशों बोला जाता ? ?
9          2  गौरव जी आप बताइए आप बीजेपी पार्टी क्यों ज्वाइन...
10         2  भारत राजधानी दिल्ली पूर्ण राज्य दर्जा मिलना चा...
11         2  हमारे भारत कुछ जवान शहीद गए आप लोग अपनी अपनी र...
12         2            अधिकांशतः भ्रष्ट नेता इंडिया क्यों हैं?
13         2  जम्मू कश्मीर पुलवामा आतंकी हमले पाकिस्तान प्रत...
14         2  आपके अनुसार भारतीय राजनीति किस बड़े सुधार आवश्...
15         2  कश्मीर कुछ लोग पाकिस्तान साथ देते अब भारत उनके...
16         2     भारत पाकिस्तान चहिते उनको सरकार कुछ सजा देगी ?
17         2                 100 प्रतिशत मुस्लिम क्यों आतंकवादी
18         2  पुलवामा आतंकवादी हमले दिन बीत गए तक सरकार आतंक...
19         2  आजकल भारत देश बहुत सारे मुद्दे उसमें प्रमुख मु...
20         2  भारत कुछ जगहों कश्मीरियों निकाला क्योंकि कुछ क...
21         2  लोकसभा चुनाव मद्देनजर भाजपा एआईडीएमके गठबंधन आ...
22         2  भारत कानून वह लचीला ? अगर हां क्यों है? ? विस्...
23         2  भारत लोगों पुलिस साथ अच्छे संबंध चलते ? इसका म...
24         2      कुछ बदनाम किस्से सऊदी रॉयल फैमिली बारे बताएं?
25         2  अगर स्मृति ईरानी राहुल गांधी खिलाफ 2019 चुनाव ...
26         2  निरंतर रही फौजियों मौत लेकर क्या सरकार बड़ा कद...
27         2  भारत युवा वर्ग आज किस दिशा ! आपके हिसाब बताइए ...
28         2               क्या बेरोजगार चुनावी मुद्दा चुका है?
29         2  हम हमारे देश अर्थव्यवस्था किस प्रकार मजबूत सकत...
..       ...                                                ...
345        3                 भारत सबसे ज्यादा पापुलर हीरो कोन ?
346        3                    टोटल धमाल ट्रेलर आपको कैसा लगा?
347        3                    टाइगर श्रॉफ पहली फिल्म क्या नाम
348        3                    नीता बच्चन करोड़पति कैसे बने? ?
349        3               धर्मेंद्र सबसे पहली मूवी कौनसी है? ?
350        3                    सैफ अली खान पहली पत्नी नाम क्या
351        3             करिश्मा कपूर सबसे पहला पिक्चर कौन था ?
352        3                    भारत कौन गाने ऑस्कर नवाजा है? ?
353        3                राजा हिंदुस्तानी मूवी हीरोइन नाम? ?
354        3             धडकन मूवी सबसे बेस्ट सॉन्ग कोनसा है? ?
355        3                        मोनी रॉय पहली फिल्म कौन है?
356        3                 बॉलीवुड सबसे बोल्ड अभिनेत्री कौन ?
357        3  फिल्म स्टार सलमान खान सुपरस्टार हीरो बोलता फुल...
358        3                           बार कौन फिल्म हिट होगी ?
359        3                          सलमान खान पहली मूवी कौन ?
360        3  क्या उरी मूवी 200 क्लब शामिल पाएगी विचार व्यक्...
361        3            तेरी गलियों रखेंगे कदम कौन मूवी सॉन्ग ?
362        3                  मटक कली मटक कली कौन सा मूवी सॉन्ग
363        3           आजकल हिन्दी फिल्मे ठीक पुरानी फील्मे ठीक
364        3                                  सिंबा मूवी कैसी ?
365        3         सबसे ज्यादा करो रुपए कमाने वाली मूवी कौन ?
366        3                 सबसे बढ़िया पिक्चर कौन लगी थिएटर ?
367        3       क्या साल तीनों खानों पिक्चरें फ्रॉक जाएंगे ?
368        3  ऐसी मूवी नाम बताओ जिसमे अमिताभ बच्चन लोगों साम...
369        3                   तेरे नाम कौन सा हीरो पाया जाता ?
370        3                    आपके सबसे अच्छे अभिनेता कोनसे ?
371        3             शाहरुख खान जीरो सिनेमा डायरेक्टर कौन ?
372        3                               इज किंग ऑफ बॉलीवुड ?
373        3                   बॉलीवुड किसके पास संपत्ति ज्यादा
374        3              रणबीर सिंह सिंबा आपको क्यों पसंद आए ?

[375 rows x 2 columns]

### split the dataset into train & validation sets

In [0]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size=0.2)

In [0]:
df1=train.reset_index()
df2=test.reset_index()

In [0]:
train_df=df1.drop('index',axis=1)
valid_df=df2.drop('index', axis=1)

In [0]:
path = '/content'

## language model

In [0]:
# Language model data
data_lm = TextLMDataBunch.from_df(path,train_df , valid_df )

# Classifier model data
data_clas = TextClasDataBunch.from_df(path,train_df,valid_df , vocab=data_lm.train_ds.vocab, bs=32)

# Fine-Tuning the Pre-Trained Model and Making Predictions

In [0]:
learn = language_model_learner(data_lm,arch=AWD_LSTM,pretrained=False, drop_mult=0.3)

In [33]:
# train the learner object with learning rate = 1e-2
learn.fit_one_cycle(5, 1e-2)

In [0]:
learn.save_encoder('ft_enc')

In [0]:
learn = text_classifier_learner(data_clas,arch=AWD_LSTM,pretrained=False, drop_mult=0.3)
learn.load_encoder('ft_enc')

In [36]:
learn.fit_one_cycle(5, 1e-2)

## predictions

In [37]:
# get predictions
preds, targets = learn.get_preds()

predictions = np.argmax(preds, axis = 1)
pd.crosstab(predictions, targets)

col_0   0   1
row_0        
0      58  17

In [38]:
valid_df

Category                                              Query
0         3             शाहरुख खान जीरो सिनेमा डायरेक्टर कौन ?
1         2                          आज समय किस राजनीति होती ?
2         2  कांग्रेस सरकार सालो जितना देश पैसा बर्बाद हुआ ...
3         2                        एमएलए फुल फॉर्म क्या होती ?
4         2  अगर भाजपा 2019 चुनाव हार जाती भाजपा समर्थक इसक...
5         2  जम्मू कश्मीर महबूबा मुफ्ती क्या कांग्रेस साथ ग...
6         2  क्या पाकिस्तान सबक सिखना जरूरी ताकि भविष्य वह ...
7         2  नरेंद्र मोदी बराबर नेता हमारे देश अपनी राय दीज...
8         2                        2019 देश अगला पीएम कौन होगा
9         2  बात कितनी सच्चाई दिल्ली मुख्यमंत्री अरविंद केज...
10        2  Rajasthan ke मुख्यमंत्री रूप आप किसे चुनना चाह...
11        2  क्या नरेंद्र मोदी जी लोकप्रियता कमी आई 2014 वा...
12        2               आज भारत वर्ष सबसे ज्वलंत मुद्दा क्या
13        2      जनकल्याण हिसाब भारत केसी राजनीति होनी चाहिए ?
14        2  भाइयो मुझे बताइये कौन कौन प्रदेश भाजपा सरकार ब...
15        2           मध्य प्रदेश अगला मुख्यमंत्री कौन रहेगा ?
16        3                               सलमान खान इज बेस्ट ?
17        3                    टोटल धमाल ट्रेलर आपको कैसा लगा?
18        2                   क्या धर्म राजनीति करना जरूरी है?
19        3                    सैफ अली खान पहली पत्नी नाम क्या
20        2  आपकी नज़रों आज तक सबसे सफल प्रधानमंत्री कौन क्...
21        2  देश नंबर वन कैसे जाता क्या वहां लोग नंबर वन पै...
22        2    क्या बीजेपी सरकार लिए राम मंदिर चुनावी मुद्दा ?
23        2          2019 चुनाव बाद भारत मुख्यमंत्री कौन बनेगा
24        2  भारत युवा वर्ग आज किस दिशा ! आपके हिसाब बताइए ...
25        2               क्या बेरोजगार चुनावी मुद्दा चुका है?
26        2                   क्या आरक्षण जाति आधार होना सही ?
27        2  सरदार वल्लभ भाई पटेल आरएसएस क्यों बैन था इसके ...
28        2  क्या मोदी जी 2019 चुनाव जीत पाएंगे कृपया 2004 ...
29        2              भारत पीएम कौन होगा राहुल गांधी मोदी ?
..      ...                                                ...
45        2     भारत पाकिस्तान चहिते उनको सरकार कुछ सजा देगी ?
46        3                              मन दिमाग़ क्या अंतर ?
47        2  पार्टियां धर्म जाति नाम राजनीति लोगों भड़काने क...
48        2  करतारपुर कॉरिडोर न्यू रखने दौरान पाकिस्तान प्र...
49        2  क्या आपको कम्युनिस्ट विचारधारा अच्छी लगती अगर ...
50        3  कौनसी फिल्म किताब आपके जीवन हमेशा लिए बदल सकती...
51        2                                    सबसे बुरा कौन ?
52        2  कांग्रेस मध्यप्रदेश ज्योतिराज सिंधिया कमलनाथ अ...
53        2                                     बुदनी जीतेगा ?
54        3  मलाइका अरोरा अर्जुन कपूर रिलेशनशिप बारे आपका क...
55        2            राष्ट्रीय लोकतांत्रिक पार्टी किसने बनाई
56        2  जम्मू कश्मीर पुलवामा आतंकी हमले पाकिस्तान प्रत...
57        2  मोदी जी ओर राहुल गांधी आप किसे समर्थन दोगे ओर ...
58        2  आप हमारे पंतप्रधान नरेंद्र मोदी जी क्यों सपोर्...
59        2  कश्मीरी हिन्दुओ कत्लेआम क्या कार्यवाही हुई कित...
60        3       क्या साल तीनों खानों पिक्चरें फ्रॉक जाएंगे ?
61        2         जमात-ए-इस्लामी ठिकानों सरकार क्यों सील है?
62        2  क्या नरेंद्र मोदी सच चाचा नेहरू वजह देश प्रधान...
63        2            अधिकांशतः भ्रष्ट नेता इंडिया क्यों हैं?
64        3                    रणवीर सिंह शादी किसके साथ हुई ?
65        3  लव यात्री मूवी बॉक्स ऑफिस अच्छा क्यों कमा पा र...
66        2             राजस्थान बार किसकी सरकार बनेगी दम बताओ
67        2  वोटरों तक पहुंच बनाने राजनीतिक दलों लिए क्या फ...
68        2                मध्यप्रदेश 2019 किसकी सरकार बनेगी ?
69        3   बजरंगी भाईजान मूवी सलमान खान क्या नाम बताते थे ?
70        2  अमेरिका राष्ट्रपति डोनाल्ड ट्रंप पाकिस्तान बार...
71        3                            सालमान खान पीता कोन हे?
72        2  चुनाव समय प्रत्याशी वोट मगाने आयें तों हमे किस...
73        2                         राजस्थान किसकी सरकार बनेगी
74        3  रणवीर सिंह दीपिका पादुकोण क्या इंडिया शादी करन...

[75 rows x 2 columns]

In [39]:
learn.predict("रणवीर सिंह की शादी किसके साथ हुई है ? ")

(Category 2, tensor(0), tensor([0.8119, 0.1881]))

# Documentation:-

# Step by step explanaion:-

- Import Necessary Libraries 
- Imported Dataset of queries and stop words.
- Reindexed and renamed column by removing unnecessary columns to category and query.
```
df = df.drop(df[['Unnamed: 2','Unnamed: 3','Unnamed: 4','Unnamed: 5','Unnamed: 6']]
```

- And then calculated total no. of missing values and removed all of them.
```
df=df.dropna(subset = ['Category', 'Query'])
```
- Then I tokenized all the query to remove all the stop words.
  The first step of processing we make the texts go through is to split the raw sentences into words, or more exactly tokens. The easiest way to do this would be to split the string on spaces, but we can be smarter:

   - we need to take care of punctuation
   - some words are contractions of two different words, like isn't or don't
   
- Now spliting the dataset to build the model and so I splitted into 80:20 where 80% is training data and 20% is testing data.

-  So i have used **TextLMDataBunch** Language model and **LSTM** (Long Short Term Memory)
Because it’s really hard for a model to learn language when only provided with a single value — the sentiment — FastAI lets you first train a language model — a model that predicts the next word — and then use that encoder for the model that actually classifies the sentiment.

This allows the classification model to focus on learning the sentiment instead of having to learn the whole language from scratch. This approach offers astonishingly good results.

```
    **#Language model data**
    data_lm = TextLMDataBunch.from_df(path,train_df , valid_df )

    **#Classifier model data**
    data_clas = TextClasDataBunch.from_df(path,train_df,valid_df , vocab=data_lm.train_ds.vocab, bs=32)
```

-  And finally we get the 77% of accuracy.
![Imgur](https://i.imgur.com/CE8IBRj.png)


# - **Thank You !**